File: paper-folding-curve.py
import sys
import math
import argparse
from time import sleep

class CustomFormatter(argparse.ArgumentDefaultsHelpFormatter, argparse.RawDescriptionHelpFormatter):
    pass

# Print iterations progress
def printProgressBar (iteration, total, prefix = '', suffix = '', decimals = 1, length = 100, fill = 'X'):
    """
    Call in a loop to create terminal progress bar
    @params:
        iteration   - Required  : current iteration (Int)
        total       - Required  : total iterations (Int)
        prefix      - Optional  : prefix string (Str)
        suffix      - Optional  : suffix string (Str)
        decimals    - Optional  : positive number of decimals in percent complete (Int)
        length      - Optional  : character length of bar (Int)
        fill        - Optional  : bar fill character (Str)
    """
    percent = ("{0:." + str(decimals) + "f}").format(100 * (iteration / float(total)))
    filledLength = int(length * iteration // total)
    bar = fill * filledLength + '-' * (length - filledLength)
    sys.stdout.write('\r%s |%s| %s%% %s \r' % (prefix, bar, percent, suffix))
    sys.stdout.flush()
    #print('\r%s |%s| %s%% %s' % (prefix, bar, percent, suffix), end = '\r')
    # Print New Line on Complete
    if iteration == total: 
        print()

  
global compteurProgressBar
compteurProgressBar=0

# Pour compter le nombre de virage total
def bin(order):
    c=0
    out=0
    while c < order:
        pas=pow(2, c)
        out+=pas
        c+=1
    return out

def path(order, s, e, file):
    start=s+1
    current=s+1
    end=e
    modulo=2
    if (file!=""):
        printProgressBar(0, end-start, prefix = 'Progress:', suffix = 'Complete  ', length = 50)
    while s <= e:
        curve(order, start, current, end, modulo, file)
        current+=1
        s+=1
    #sys.stdout.write("\n")
    
def curve(order, start, current, end, modulo, file):
    global compteurProgressBar
    if current<=end+1:
        if (current % modulo):
            value = ("0", "1")[int(math.ceil(current / modulo)) % 2]
            if (file==""):
                sys.stdout.write(value)
                sys.stdout.flush()
            else:
                printProgressBar(compteurProgressBar-1, end-start, prefix = 'Progress:', suffix = 'Complete  ', length = 50)
                compteurProgressBar+=1
                file.write(value)     
        else:
            curve(order-1, start, current, end, modulo*2, file)


# Creating a parser
tab = "\t\t\t"
e = 'Get a stripe of paper.\nFold it into half.\n'
e += 'Fold it into half again in the same direction.\n'
e += 'Fold it the third time into half, in the same direction.\n'
e += 'Then open it, so that every fold is at a right angle.\n'
e += 'Viewing it from the edge you could see a pattern like this:\n'
e += '   _ \n'
e += '|_| |_\n'
e += '     _|\n'
e += "\n"
e += "Example:\n"
e += '--order 3 --start 3 --end 6\n'
e += '--order 5 --start 16 --end 18\n'
e += '--order 10 --start 999 --end 1011\n'
e += '--order 20 --start 100 --end 111\n'
e += '--order 50 --start 562949953421300 --end 562949953421400\n'
e += "\n"
e += "Output:\n"
e += '1100\n'
e += '110\n'
e += '1110010001101\n'
e += '100111001000\n'
e += '10001100100111011001110010011101100011001001110110011100100011011000110010011101100111001001110110001\n'

t = "Paper-folding curve.\n"
t += "Coding Games\n"
t += "https://www.codingame.com/training/community/paper-folding-curve\n"
t += "create by java_coffee_cup\n"
t += "resolved by Xeno_PXD\n"


parser = argparse.ArgumentParser(description=t, 
                                 epilog=e, formatter_class=CustomFormatter)

# Adding arguments
parser.add_argument('-o', '--order', dest='order', nargs=1, type=int, help='An integer N for the order of the curve')
parser.add_argument('-s', '--start', dest='s', metavar=('START'), nargs=1, type=int, help='A starting index')
parser.add_argument('-e', '--end', dest='e', metavar=('END'), nargs=1, type=int, help='A ending index')
parser.add_argument('-out', '--output', dest='output', metavar=('OUTPUT'), nargs=1, type=str, help='output file')
parser.add_argument('-n', '--number', dest='number', action='store_true', help='return number curve')

# Parsing arguments
args = parser.parse_args()
nb=bin(args.order[0])
start=0
end=0

if (args.number):
    print(nb)
    exit()
    
if (args.e is None):
    end = nb
else:
    if args.e[0]>nb:
        end=nb
    else:
        end=args.e[0]
        
if (args.s is None):
    start=0
else:
    if args.s[0]<0:
        start=0
    else:
        start=args.s[0]

file=""
if  (args.output is None):
    file=""
else:
    file = open(args.output[0], 'w+')

#print(str(args.order) +" "+ str(args.s) +" "+ str(args.e))
path(args.order[0], start, end, file)

File: paper-folding-curve.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <math.h>
#include <cstdlib>
#include <map>
#include <algorithm>
#include <stdlib.h>
#include <fstream>
#include <time.h>

using namespace std;

// g++ -std=c++11 paper-folding-curve.cpp -o paper-folding-curve

ofstream myfile;
bool file = false;
long compteurProgressBar = 0;

class CLParser
{
public:

    CLParser(int argc_, char * argv_[],bool switches_on_=false);
    ~CLParser(){}

    string get_arg(int i);
    string get_arg(string s);
    bool isExist(string s);

private:

    int argc;
    vector<string> argv;

    bool switches_on;
    map<string,string> switch_map;
};

CLParser::CLParser(int argc_, char * argv_[],bool switches_on_)
{
    argc=argc_;
    argv.resize(argc);
    copy(argv_,argv_+argc,argv.begin());
    switches_on=switches_on_;

    //map the switches to the actual
    //arguments if necessary
    if (switches_on)
    {
        vector<string>::iterator it1,it2;
        it1=argv.begin();
        it2=it1+1;

        while (true)
        {
            if (it1==argv.end()) break;
            if (it2==argv.end()) break;

            if ((*it1)[0]=='-')
                switch_map[*it1]=*(it2);

            it1++;
            it2++;
        }
    }
}

string CLParser::get_arg(int i)
{
    if (i>=0&&i<argc)
        return argv[i];

    return "";
}

string CLParser::get_arg(string s)
{
    if (!switches_on) return "";

    if (switch_map.find(s)!=switch_map.end())
        return switch_map[s];

    return "";
}

bool CLParser::isExist(string s)
{
    if (switches_on) return true;

    //if (switch_map.find(s)!=switch_map.end())
    //    return true;

    return false;
}

long bin(int order) {
    long c=0;
    long out=0;
    while(c < order) {
      out+=pow(2, c);
      c+=1;
    }
    return out;
}

void curve(int order, long start, long current, long end, long modulo) {
    if (current<=end+1) {
        if (current % modulo) {
          compteurProgressBar++;
          string value = "1";
          if (long(ceil(current / modulo)) % 2) {
            value = "0";
          }
          if (file) {
            myfile << value;
            //string progress = "\r5";
            //cout << "\r";
            //cout << compteurProgressBar << " \r";
            //std::cout.flush();

          } else {
            cout << value;
          }
        } else {
           curve(order-1, start, current, end, modulo*2);
        }
    }
}

void path(int order, long s, long e) {
    long start=s+1;
    long current=s+1;
    long end=e;
    long modulo=2;
    if (file) {
      cout << (end-start)+1;
      cout << " curves in progress..." << endl;
      //cout << "0" << endl;
    }
    while(s <= e) {
      curve(order, start, current, end, modulo);
      current+=1;
      s+=1;
    }
    if(!file) cout << endl;
}

int main(int argc, char* argv[]) {

    CLParser cmd_line(argc,argv,true);

        //int order = atoi(argv[1]);
    //nb=bin(order);
    //cout << nb << endl;

    string temp;
    bool tempExist;

    int order;
    temp=cmd_line.get_arg("--order");
    if (temp!="") {
      order = stoi(temp);
    }

    long nb;
    nb=bin(order);
    //cout << nb << endl;
/*
    tempExist=cmd_line.isExist("--number");
    cout << tempExist << endl;
    if (tempExist) {
      cout << nb << endl;
    }
*/
    long start;
    temp=cmd_line.get_arg("--start");
    if (temp!="") {
      start = stol(temp);
      if (start<0) start=0;
    } else {
      start = 0;
    }

    long end;
    temp=cmd_line.get_arg("--end");
    if (temp!="") {
      end = stol(temp);
      if (end>nb) end=nb;
    } else {
      end = nb;
    }

    temp=cmd_line.get_arg("--output");
    if (temp!="") {
      myfile.open (temp);
      file = true;
    }

    clock_t tStart = clock();
    /* Do your stuff here */

    path(order, start, end);

    if(file) {
      myfile.close();
      printf("Time taken: %.3fs\n", (double)(clock() - tStart)/CLOCKS_PER_SEC);
    }
    return 0;
}
